src/java/org/apache/cassandra/index/sai/disk/v1/bbtree/BlockBalancedTreeWalker.java

```diff
@@ -0,0 +1,309 @@
+        }
+
+        public void pop()
+        {
+            nodeID /= 2;
+            level--;
+            leafNode = false;
+        }
+
+        public boolean isLeafNode()
+        {
+            return nodeID >= numLeaves;
+        }
+
+        public boolean nodeExists()
+        {
+            return nodeID - numLeaves < numLeaves;
+        }
+
+        public int getNodeID()
+        {
+            return nodeID;
+        }
+
+        public byte[] getSplitPackedValue()
+        {
+            assert !isLeafNode();
+            assert splitPackedValueStack[level] != null : "level=" + level;
+            return splitPackedValueStack[level];
+        }
+
+        public long getLeafBlockFP()
+        {
+            assert isLeafNode() : "nodeID=" + nodeID + " is not a leaf";
+            return leafBlockFPStack[level];
+        }
+
+        public byte[] getSplitValue()
+        {
+            assert !isLeafNode();
+            return splitValuesStack[level];
+        }
+
+        private void readNodeData(boolean isLeft)
+        {
+            leafBlockFPStack[level] = leafBlockFPStack[level - 1];
+
+            // read leaf block FP delta
+            if (!isLeft)
+                leafBlockFPStack[level] += in.readVLong();
+
+            leafNode = isLeafNode();
+            if (!leafNode)
+            {
+                // read prefix, firstDiffByteDelta encoded as int:
+                int code = in.readVInt();
+                int prefix = code % (1 + bytesPerValue);
+                int suffix = bytesPerValue - prefix;
+
+                if (splitValuesStack[level] == null)
+                {
+                    splitValuesStack[level] = new byte[bytesPerValue];
+                }
+                System.arraycopy(splitValuesStack[level - 1], 0, splitValuesStack[level], 0, bytesPerValue);
+                if (suffix > 0)
+                {
+                    int firstDiffByteDelta = code / (1 + bytesPerValue);
+                    if (negativeDeltas[level])
+                        firstDiffByteDelta = -firstDiffByteDelta;
+                    int oldByte = splitValuesStack[level][prefix] & 0xFF;
+                    splitValuesStack[level][prefix] = (byte) (oldByte + firstDiffByteDelta);
+                    in.readBytes(splitValuesStack[level], prefix + 1, suffix - 1);
+                }
+
+                int leftNumBytes = nodeID * 2 < numLeaves ? in.readVInt() : 0;
+
+                leftNodePositions[level] = in.getPosition();
+                rightNodePositions[level] = leftNodePositions[level] + leftNumBytes;
+            }
+        }
+
+        private int getTreeDepth()
+        {
+            // First +1 because all the non-leave nodes makes another power
+            // of 2; e.g. to have a fully balanced tree with 4 leaves you
+            // need a depth=3 tree:
+
+            // Second +1 because MathUtil.log computes floor of the logarithm; e.g.
+            // with 5 leaves you need a depth=4 tree:
```

Comment:
If the number of leaves is a power of two then `MathUtil.log` won't apply the flooring and we will needlessly add one. I understand what we want here is to ceil the log, instead of flooring. 

I think we can achieve that without resorting to `Math.ceil(Math.log(n) / Math.log(2))` with Guava's
```java
IntMath.log2(numLeaves, RoundingMode.CEILING)
```
Or we can skip the call to Guava and just use:
```java
32 - Integer.numberOfLeadingZeros(x - 1)
```

Reply:
Firstly, neither of the above suggestions work correctly because they underestimate the tree depth for low values of `numLeaves`. Secondly, I don't believe that this method was ever meant to be 100% accurate for the tree depth, it returns a value that is as close as possible but will have a tendency to overestimate.

If we can come up with a method that is 100% accurate then I am more than happy to change the method but for the time being I'd prefer to leave as is.

I am thinking that we could just add the tree depth to the on-disk format because the write side must know the value.

Reply:
> I am thinking that we could just add the tree depth to the on-disk format because the write side must know the value.

That doesn't sound like a bad idea.

Reply:
I have ended up adding the `treeDepth` to the packed index and removing this method.